stub_uploader/get_version.py

```diff
@@ -83,28 +60,90 @@ def check_exists(distribution: str) -> bool:
     raise ValueError("Error while verifying existence")
 
 
-def main(typeshed_dir: str, distribution: str, version: Optional[str]) -> int:
-    """A simple function to get version increment of a third-party stub package.
-
-    Supports basic reties and timeouts (as module constants).
-    """
-    pypi_versions = fetch_pypi_versions(distribution)
-    if not version:
-        # Use the METADATA.toml version, if not given one.
-        version = read_base_version(typeshed_dir, distribution)
-    matching = [v for v in pypi_versions if v.startswith(f"{version}.")]
-    if not matching:
-        return -1
-    increment = max(int(v.split(".")[-1]) for v in matching)
-    return increment
-
-
-if __name__ == "__main__":
-    parser = argparse.ArgumentParser()
-    parser.add_argument("typeshed_dir", help="Path to typeshed checkout directory")
-    parser.add_argument("distribution", help="Third-party distribution to build")
-    parser.add_argument(
-        "version", nargs="?", help="Base version for which to get increment"
+def extend_to_specificity(ver: list[int], specificity: int) -> None:
+    ver.extend([0] * (specificity - len(ver)))
+
+
+def compute_incremented_version(
+    version_spec: str, published_versions: Union[list[str], list[Version]]
+) -> Version:
+    # The most important postcondition is that the incremented version is greater than
+    # all published versions. This ensures that users who don't pin get the most
+    # up to date stub. If we ever maintain multiple versions for a stub, this will
+    # need revisiting.
+    max_published = max(
+        ((Version(v) if isinstance(v, str) else v) for v in published_versions),
+        default=Version("0"),
+    )
+
+    # The second thing we try to do (but don't guarantee), is that the incremented
+    # version will satisfy the version_spec (defined precisely by the `compatible`
+    # specifier below). This allows users to have expectations of what a stub package
+    # will contain based on the upstream version they're targeting.
+    if version_spec.endswith(".*"):
+        compatible = SpecifierSet(f"=={version_spec}")
+    else:
+        compatible = SpecifierSet(f"=={version_spec}.*")
+
+    # See what the base version and specificity we used in METADATA.toml
+    version_base = Version(version_spec.removesuffix(".*"))
+    specificity = len(version_base.release)
+
+    if max_published.epoch > 0 or version_base.epoch > 0:
+        raise NotImplementedError("Epochs in versions are not supported")
+
+    # We'll try to bump the fourth part of the release. So e.g. if our version_spec is
+    # 1.1, we'll release 1.1.0.0, 1.1.0.1, 1.1.0.2, ...
+    # But if our version_spec is 5.6.7.8, we'll release 5.6.7.8.0, 5.6.7.8.1, ...
```

Comment:
Why change this behavior? For example, types-tzlocal currently has version 4.2.2, while tzlocal has version 4.2.

Reply:
I think it helps disambiguate "upstream version" from "types version". I've seen someone be confused by this and pin unnecessarily strictly.

E.g. if tzlocal released 4.2.1 that would still be compatible with 4.2 specified in METADATA.toml. But the correct corresponding version of types-tzlocal is still 4.2.2.

The issue comes about because the target version in METADATA.toml is hidden information to the average user. Very few packages use the fourth spot, so defaulting to the fourth spot in practice makes it slightly clearer that the version isn't 1:1 with upstream versions.

Reply:
I'm fine, but I'm concerned this could break types-tzlocal (and potentially others). I would like to see this changed in a separate PR to keep possible breakage per PR minimal.

Reply:
I'm curious if your concern can be turned into suggestions for tests. I just added a `test_version_increment` that runs `determine_incremented_version` on all distributions we publish. This checks that all invariants we assert are met for all currently distributed packages.

But that aside, I can comment this change out for now.